Segmentation violation when adding txn with gRPC client

## [Segmentation violation when adding txn with gRPC client]

### Description

Hi, I am currently working on a TypeScript gRPC client for this SDK (available [here](https://github.com/PtitLuca/0xpolygon-ts)).

I was adding the support for the Txn Pool Operator, but when I was testing my addTxn function a SIGSEGV happened.

### Your environment

* **OS**: Pop OS 21.04
* **Polygon SDK Version**: 0.1.0
* **Branch**: default

### Steps to reproduce

First, I started a dev server using the command provided [here](https://sdk-docs.polygon.technology/docs/how-tos/howto-query-json-rpc).

```go
go run main.go genesis --premine 0x1010101010101010101010101010101010101010
go run main.go server --chain genesis.json --dev --log-level debug --grpc :9632
```

I generated the protobuf files for my client using the following command and the `txpool/proto/operator.proto` file of the polygon-sdk repository on branch develop:
```sh
yarn proto-loader-gen-types --grpcLib=@grpc/grpc-js --outDir=src/proto proto/*.proto
```

Then using my client I ran the following test with the command `yarn test`:

```js
const client = new PolygonTxnPoolOperatorClient('localhost', 9632);

await client.addTxn(null, '0x1010101010101010101010101010101010101010');
```

where `PolygonTxnPoolOperatorClient` is the following:

```ts
class PolygonTxnPoolOperatorClient {
    private readonly address: string;
    private client: TxnPoolOperatorClient;

    constructor(hostname: string, port: number) {
        this.address = `${hostname}:${port}`;
	this.client = new grpcObj.v1.TxnPoolOperator(this.address, credentials.createInsecure());
    }

    addTxn(raw: Any | null, from: string): Promise<Empty__Output> {
        return new Promise((resolve, reject) => {
            const addTxnReq: AddTxnReq = {
                raw,
		from,
            };

	    this.client.addTxn(addTxnReq, (error?: ServiceError, result?: Empty__Output) => {
	        if (error) {
		    return reject(error);
		}
	        if (result) {
		    return resolve(result);
	        }
		throw new Error('unknown error: no error and status could be retrieved from the request');
	    });
        });
    }
}
```

### Expected behaviour

I was actually trying to understand the behaviour of this method ^^.

### Actual behaviour

The polygon server crashed because of a segmentation violation.

### Logs

```sh
➜  polygon-sdk git:(develop) ✗ go run main.go server --chain genesis.json --dev --log-level debug --grpc :9632
2021-08-01T20:11:06.750+0200 [INFO]  polygon: Data dir: path=./test-chain
2021-08-01T20:11:06.751+0200 [INFO]  polygon.network: LibP2P server running: addr=/ip4/127.0.0.1/tcp/43345/p2p/16Uiu2HAmF3obRnYGsdzThy2irJRAD3DoEiDgsrHbJR8QnhGKVW7x
2021-08-01T20:11:06.759+0200 [INFO]  polygon.blockchain: Current header: hash=0xb66970dc536ae4a18cc6c9b760f17afa6e31cf03d579a77cbf86ce276e8f602c number=0
2021-08-01T20:11:06.759+0200 [INFO]  polygon.blockchain: genesis: hash=0xb66970dc536ae4a18cc6c9b760f17afa6e31cf03d579a77cbf86ce276e8f602c
2021-08-01T20:11:06.759+0200 [INFO]  polygon: GRPC server running: addr=127.0.0.1:9632
2021-08-01T20:11:06.759+0200 [INFO]  polygon.jsonrpc: http server started: addr=127.0.0.1:8545
2021-08-01T20:11:06.759+0200 [INFO]  polygon.consensus.dev: consensus started
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0xc68455]

goroutine 182 [running]:
github.com/0xPolygon/minimal/txpool.(*TxPool).AddTxn(0xc000596230, 0x1246360, 0xc0016f8c90, 0xc0016f5bc0, 0xc000596230, 0xc0016f8c90, 0xc0004c5ba0)
        /home/ptitluca/GolandProjects/polygon-sdk/txpool/operator.go:23 +0x55
github.com/0xPolygon/minimal/txpool/proto._TxnPoolOperator_AddTxn_Handler(0xec1180, 0xc000596230, 0x1246360, 0xc0016f8c90, 0xc0016fac60, 0x0, 0x1246360, 0xc0016f8c90, 0xc00003a5a0, 0x2c)
        /home/ptitluca/GolandProjects/polygon-sdk/txpool/proto/operator_grpc.pb.go:151 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00033e540, 0x124e798, 0xc000a70780, 0xc0004b2100, 0xc000516660, 0x1a37db8, 0x0, 0x0, 0x0)
        /home/ptitluca/GolandProjects/polygon-sdk/vendor/google.golang.org/grpc/server.go:1217 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc00033e540, 0x124e798, 0xc000a70780, 0xc0004b2100, 0x0)
        /home/ptitluca/GolandProjects/polygon-sdk/vendor/google.golang.org/grpc/server.go:1540 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc001f20720, 0xc00033e540, 0x124e798, 0xc000a70780, 0xc0004b2100)
        /home/ptitluca/GolandProjects/polygon-sdk/vendor/google.golang.org/grpc/server.go:878 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/ptitluca/GolandProjects/polygon-sdk/vendor/google.golang.org/grpc/server.go:876 +0x1fd
exit status 2
```

### Proposed solution

The error that caused the segmentation violation is self explaining : I sent a null raw field in the transaction and the server is trying to access it's content in the `txpool/operator.go` file, in the function `AddTxn` on line 23 :

```go
// AddTxn adds a local transaction to the pool
21 func (t *TxPool) AddTxn(ctx context.Context, raw *proto.AddTxnReq) (*empty.Empty, error) {
22	txn := new(types.Transaction)
23	if err := txn.UnmarshalRLP(raw.Raw.Value); err != nil {
24		return nil, err
25	}
26
	if raw.From != "" {
		from := types.Address{}
		if err := from.UnmarshalText([]byte(raw.From)); err != nil {
			return nil, err
		}
		txn.From = from
	}

	if err := t.AddTx(txn); err != nil {
		return nil, err
	}

	return &empty.Empty{}, nil
}
```

Should you check the content of the `raw` parameter, and return an error if it is actually nil ?
Like this:

```go
// AddTxn adds a local transaction to the pool
func (t *TxPool) AddTxn(ctx context.Context, raw *proto.AddTxnReq) (*empty.Empty, error) {
	if raw.Raw == nil {
		return nil, fmt.Errorf("transaction's field raw is empty")
	}

	txn := new(types.Transaction)
...
```

--- a/txpool/operator.go
+++ b/txpool/operator.go
@@ -2,6 +2,7 @@ package txpool
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/0xPolygon/minimal/txpool/proto"
 	"github.com/0xPolygon/minimal/types"
@@ -19,6 +20,10 @@ func (t *TxPool) Status(ctx context.Context, req *empty.Empty) (*proto.TxnPoolSt
 
 // AddTxn adds a local transaction to the pool
 func (t *TxPool) AddTxn(ctx context.Context, raw *proto.AddTxnReq) (*empty.Empty, error) {
+	if raw.Raw == nil {
+		return nil, fmt.Errorf("transaction's field raw is empty")
+	}
+
 	txn := new(types.Transaction)
 	if err := txn.UnmarshalRLP(raw.Raw.Value); err != nil {
 		return nil, err

--- a/txpool/txpool_test.go
+++ b/txpool/txpool_test.go
@@ -1,7 +1,9 @@
 package txpool
 
 import (
+	"context"
 	"fmt"
+	"github.com/0xPolygon/minimal/txpool/proto"
 	"math/big"
 	"strconv"
 	"testing"
@@ -218,4 +220,14 @@ func TestTxnQueue_Heap(t *testing.T) {
 			assert.Equalf(t, txns[i].GasPrice, txn.GasPrice, "Expected output mismatch")
 		}
 	})
+}
+
+func TestTxnOperatorAdd(t *testing.T) {
+	pool, err := NewTxPool(hclog.NewNullLogger(), false, &mockStore{}, nil, nil)
+	assert.NoError(t, err)
+
+	txnReq := new(proto.AddTxnReq)
+	response, err := pool.AddTxn(context.Background(), txnReq)
+	assert.Errorf(t, err, "transaction's field raw is empty")
+	assert.Nil(t, response)
 }

--- a/txpool/txpool_test.go
+++ b/txpool/txpool_test.go
@@ -222,7 +222,7 @@ func TestTxnQueue_Heap(t *testing.T) {
 	})
 }
 
-func TestTxnOperatorAdd(t *testing.T) {
+func TestTxnOperatorAddNilRaw(t *testing.T) {
 	pool, err := NewTxPool(hclog.NewNullLogger(), false, &mockStore{}, nil, nil)
 	assert.NoError(t, err)
 

--- a/txpool/txpool_test.go
+++ b/txpool/txpool_test.go
@@ -307,7 +307,7 @@ func TestTxnQueue_Heap(t *testing.T) {
 }
 
 func TestTxnOperatorAddNilRaw(t *testing.T) {
-	pool, err := NewTxPool(hclog.NewNullLogger(), false, &mockStore{}, nil, nil)
+	pool, err := NewTxPool(hclog.NewNullLogger(), false, forks.At(0), &mockStore{}, nil, nil)
 	assert.NoError(t, err)
 
 	txnReq := new(proto.AddTxnReq)

--- a/txpool/txpool_test.go
+++ b/txpool/txpool_test.go
@@ -8,11 +8,11 @@ import (
 	"strconv"
 	"testing"
 
-  "github.com/0xPolygon/minimal/chain"
+  	"github.com/0xPolygon/minimal/chain"
 	"github.com/0xPolygon/minimal/crypto"
 	"github.com/0xPolygon/minimal/helper/tests"
 	"github.com/0xPolygon/minimal/network"
-  "github.com/0xPolygon/minimal/txpool/proto"
+  	"github.com/0xPolygon/minimal/txpool/proto"
 	"github.com/0xPolygon/minimal/types"
 	"github.com/hashicorp/go-hclog"
 	"github.com/stretchr/testify/assert"

